src/coreclr/vm/siginfo.cpp

```diff
@@ -3908,7 +3952,59 @@ MetaSig::CompareElementType(
                         pOtherModule,
                         tkOther,
                         ClassLoader::ReturnNullIfNotFound,
-                        ClassLoader::FailIfUninstDefOrRef);
+                        ClassLoader::PermitUninstDefOrRef);
+
+                    // If the return type is a generic type definition, we need to
+                    // compare it to the typical method table of the internal type.
+                    if (!hOtherType.IsNull() && hOtherType.IsGenericTypeDefinition())
+                    {
+                        hInternal = TypeHandle{ hInternal.AsMethodTable()->GetTypicalMethodTable() };
+                    }
+
+                    return (hInternal == hOtherType);
+                }
+                case ELEMENT_TYPE_GENERICINST:
+                case ELEMENT_TYPE_BYREF:
+                case ELEMENT_TYPE_ARRAY:
+                case ELEMENT_TYPE_SZARRAY:
+                case ELEMENT_TYPE_PTR:
```

Comment:
Delete "Internal types can only correspond to types or value types." comment above?